server/src/main/java/org/opensearch/repositories/blobstore/BlobStoreRepository.java

```diff
@@ -1668,56 +1806,214 @@ private static boolean isIndexPresent(ClusterService clusterService, String inde
+                            + "but failed to clean up their index folders",
+                        metadata.name(),
+                        indexSnId
+                    ),
+                    e
+                );
+                return DeleteResult.ZERO;
+            } catch (Exception e) {
+                assert false : e;
+                logger.warn(new ParameterizedMessage("[{}] Exception during single stale index delete", metadata.name()), e);
+                return DeleteResult.ZERO;
+            } finally {
+                executeOneStaleIndexDelete(
+                    staleIndicesToDelete,
+                    remoteStoreLockManagerFactory,
+                    listener,
+                    snapshotShardPaths,
+                    idToShardInfoMap
+                );
+            }
+        }));
+    }
+
+    /**
+     * Finds and returns a list of shard paths that match the given index ID.
+     *
+     * @param indexId            The ID of the index to match
+     * @param snapshotShardPaths Map of snapshot shard paths and their metadata
+     * @return List of matching shard paths
+     */
+    private List<String> findMatchingShardPaths(String indexId, Map<String, BlobMetadata> snapshotShardPaths) {
+        return snapshotShardPaths.keySet().stream().filter(s -> s.startsWith(indexId)).collect(Collectors.toList());
+    }
+
+    /**
+     * Finds the shard path with the highest generation number from the given list of matching shard paths.
+     *
+     * @param matchingShardPaths List of shard paths that match a specific criteria
+     * @return An Optional containing the shard path with the highest generation number, or empty if the list is empty
+     */
+    private Optional<String> findHighestGenerationShardPaths(List<String> matchingShardPaths) {
+        return matchingShardPaths.stream()
+            .map(s -> s.split("\\" + SnapshotShardPaths.DELIMITER))
+            .sorted((a, b) -> Integer.parseInt(b[2]) - Integer.parseInt(a[2]))
+            .map(parts -> String.join(SnapshotShardPaths.DELIMITER, parts))
+            .findFirst();
+    }
+
+    /**
+     * Cleans up remote store locks for a given index entry.
+     *
+     * @param indexEntry                    The index entry containing the blob container
+     * @param shardInfo                     ShardInfo for the IndexId being cleaned up
+     * @param remoteStoreLockManagerFactory Factory for creating remote store lock managers
+     * @throws IOException If an I/O error occurs during the cleanup process
+     */
+    private void cleanupRemoteStoreLocks(
+        Map.Entry<String, BlobContainer> indexEntry,
+        ShardInfo shardInfo,
+        RemoteStoreLockManagerFactory remoteStoreLockManagerFactory
+    ) throws IOException {
+        if (shardInfo == null) {
+            releaseRemoteStoreLocksAndCleanup(indexEntry.getValue().children(), remoteStoreLockManagerFactory);
```

Comment:
When can it happen ? Do we need to write that we expect to use `fixed_path` here  ?

Reply:
This can happen for snapshots that pre-dates the new change. Let me add a note as comment.